### PR TITLE
add apply APO defaults button for items and collections

### LINKED ISSUE #3041: Add show page button for applying APO defaults
In testing, I realized that we never had a blue button for applying APO defaults on an item or collection show page. You have to paste the druid in to the bulk update interface. While going to the new layout, we should add this option to the Argo item and collection pages.

**Details**

**Object type**: this button should appear only on Collection and Item pages

**Position**: this button should appear on the "Actions" menu, after the "Purge" button

**State**: this button should only be active when the object is unlocked

**Label**: "Apply APO defaults"

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made? 🤔

Fixes #3041 - Add an Apply APO Defaults button that is shown for items and collections

~~HOLD Pending testing on QA~~
~~Note: Testing currently blocked by https://github.com/sul-dlss/dor-services-app/issues/3648 and https://github.com/sul-dlss/argo/issues/3185~~

## How was this change tested? 🤨

Localhost browser, and updated component and request specs

Integration tests passed on QA

⚡ ⚠ If this change updates the Argo UI, run all integration tests that use Argo and create a PR on that repo to fix anything this change breaks. ⚡

### CODE REVIEW

File: app/components/action_button.rb [hunk 1]
```diff
@@ -1,14 +1,14 @@
 # frozen_string_literal: true
 
-# Draws a blue button in the side menu
+# Draws a red button on the item detail page
 class ActionButton < ApplicationComponent
   # @param [Hash] properties the button properties
-  def initialize(label:, url:, method: nil, confirm: nil, new_page: nil, check_url: nil, disabled: nil)
+  def initialize(label:, url:, method: nil, confirm: nil, open_modal: false, check_url: nil, disabled: nil)
     @label = label
```
Comment (peetucket) [inline app/components/action_button.rb:7]: I got tripped up by this "new_page" attribute, which basically means "don't create a modal window" when clicked.  It you don't add a "new_page", then anything other than a GET method will not work, which I found confusing.  I switched the parameter to "open_modal" instead and adjusted the buttons to reflect this.

File: app/components/show/controls_component.rb [hunk 2]
```diff
@@ -40,29 +40,41 @@ def symphony_record?
...
     private
 
     def manage_release
-      render ActionButton.new(url: item_manage_release_path(pid), label: 'Manage release')
+      render ActionButton.new(url: item_manage_release_path(pid), label: 'Manage release', open_modal: true)
+    end
+
+    def apply_apo_defaults
+      return unless item? || collection?
+
```
Comment (peetucket) [inline app/components/show/controls_component.rb:53]: I added a second guard against this button being shown for APOs (you'll note that the view template itself actually doesn't render this button at all for APOs)

File: app/controllers/items_controller.rb [hunk 3]
```diff
@@ -220,7 +220,13 @@ def set_rights
...
     reindex
-    render status: :ok, plain: 'APO defaults applied.'
+    respond_to do |format|
+      if params[:bulk]
+        format.html { render status: :ok, plain: 'APO defaults applied.' }
+      else
+        format.any { redirect_to solr_document_path(params[:id]), notice: 'APO defaults applied!' }
+      end
+    end
   rescue Dor::Services::Client::UnexpectedResponse => e
```
Comment (peetucket) [inline app/controllers/items_controller.rb:230]: This method is currently only used by the bulk action -- this lets us re-purpose it for the new button.

File: app/components/action_button.rb [hunk 4]
```diff
@@ -1,14 +1,14 @@
 # frozen_string_literal: true
 
-# Draws a blue button in the side menu
+# Draws a red button on the item detail page
```
Comment (mjgiarlo) [inline app/components/action_button.rb:3]: Haha! `s/blue/red/` 😆 

Given we will be changing these buttons' colors over time, I imagine, I wonder if we can drop the color altogether.

File: app/components/action_button.rb [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
 class ActionButton < ApplicationComponent
   # @param [Hash] properties the button properties
-  def initialize(label:, url:, method: nil, confirm: nil, new_page: nil, check_url: nil, disabled: nil)
+  def initialize(label:, url:, method: nil, confirm: nil, open_modal: false, check_url: nil, disabled: nil)
```
Comment (jcoyne) [inline app/components/action_button.rb:6]: didn't this just flip the default?  If they specified nothing before, it would open a modal.  Now if you specify nothing it doesn't open a modal.  This seems like a change that could inadvertently cause problems.  Can you move it to a separate PR?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/components/action_button.rb [hunk 6]
```diff
@@ -22,7 +22,7 @@ def button_data
       if confirm
         # :confirm trumps :blacklight_modal, because :blacklight_modal would negate :confirm by firing the ajax request regardless of the user's decision
         data[:turbo_confirm] = confirm
-      elsif !new_page
+      elsif open_modal
```
Comment (mjgiarlo) [inline app/components/action_button.rb:25]: I'm not saying this is a problem, but I'm observing that the default behavior before (since `new_page` defaulted to `nil`) was to trigger this condition and now the default behavior is not to trigger this condition. Intended? Not a problem if you touched all the places that call this code and do not want to trigger this condition.

File: app/components/action_button.rb (hunk 5 above)
Comment (peetucket) [reply app/components/action_button.rb:6]: well yes it flips it, but makes the intent clearer to me (I wasted an hour trying to figure out why my POST request wasn't working for the new button)

could move to a new PR, but will then have to wait for this one to merge to move forward

Comment (mjgiarlo) [reply app/components/action_button.rb:6]: I think it's fine as long as you touched all the callers to ensure the right behavior?

File: app/components/action_button.rb (hunk 6 above)
Comment (peetucket) [reply app/components/action_button.rb:25]: There weren't that many spots that used this (all in the component itself) and I changed them.  i just want the attribute to match the intent ("new_page" currently means no modal AND only GET requests).

in my mind, showing a modal better implies a get request, so that name seems clearer to me

Comment (mjgiarlo) [reply app/components/action_button.rb:25]: WFM

File: app/components/action_button.rb (hunk 4 above)
Comment (peetucket) [reply app/components/action_button.rb:3]: Removed color

File: app/components/action_button.rb (hunk 5 above)
Comment (peetucket) [reply app/components/action_button.rb:6]: I  fixed all  of the callers (they were all in the `ControlsComponent`, which also had other changes as part of this PR anyway)

**Review by mjgiarlo — APPROVED** [review]

File: app/components/action_button.rb (hunk 5 above)
Comment (jcoyne) [reply app/components/action_button.rb:6]: What about [app/views/items/_button_link_list.html.erb](https://github.com/sul-dlss/argo/blob/f728b0922975739ede99bfcb86d7cf82d611d209/app/views/items/_button_link_list.html.erb) ?

Comment (jcoyne) [reply app/components/action_button.rb:6]: Okay 👍

**Review by jcoyne — APPROVED** [review]
